### DISCUSSION TITLE
ADIN2111 not works

### DISCUSSION THREAD

**Discussion by lmikl** [zephyrproject-rtos/zephyr#61040] [Q&A]
hello,

The driver for the ADIN2111 which was recently integrated into zephyr 3.4.0 does not work.

I am currently using my own driver and the component works well.

If I try to use the driver integrated in zephyr I just have the message (uart:~$) which appears on the console.
In debug mode under eclipse I think that the cpu does not start. it is idle.
![Capture d’écran de 2023-08-01 16-56-36](https://github.com/zephyrproject-rtos/zephyr/assets/127130921/57395ecf-5de2-4a30-bfeb-81021a81026d)

if someone can solve the problem.

**Comment by lmikl** [comment]
other information:
i used a GD32F427
my dts : 

&spi1 {
	pinctrl-0 = <&spi1_default>;
	pinctrl-names = "default";
	interrupts = < 0x24 0x0 >;
	cs-gpios = <&gpiob 12 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
	status = "okay";
	my_adin: adin2111@0{
		compatible = "adi,adin2111";
		reg = <0x0>;
		spi-max-frequency = <10000000>;
		int-gpios = <&gpioe 14 GPIO_ACTIVE_LOW>;
		reset-gpios = <&gpioe 15 GPIO_ACTIVE_LOW>;
		port1 {
			local-mac-address = [ 00 E0 22 FE DA C8 ];
		};
		port2 {
			local-mac-address = [ 00 E0 22 FE DA C8 ];
		};
		mdio: mdio {
			compatible = "adi,adin2111-mdio";
			status = "okay";
			#address-cells = <1>;
			#size-cells = <0>;
			phy@1 {
			  reg = <0x1>;
			  compatible = "adi,adin2111-phy";
			  status = "okay";
			};
			phy@2 {
			  reg = <0x2>;
			  compatible = "adi,adin2111-phy";
			  status = "okay";
			};
		};
	};
};

my prj.conf :

CONFIG_SPI=y
CONFIG_ETH_ADIN2111=y
CONFIG_ETH_ADIN2111_SPI_CFG0=n

**Comment by vChavezB** [comment]
I am not the owner of the driver but I have been working with the ADIN1110 . Can you set a breakpoint on the init function of the driver and see what is happening https://github.com/zephyrproject-rtos/zephyr/blob/d7504ab4744c5552c8f6adf04ee9d81d00c6a628/drivers/ethernet/eth_adin2111.c#L768

**Reply by lmikl** [reply]
Yes, I set a breakpoint at the init function. but I never go there.
when I start the project in debug mode and, after a few moments, I pause it, I arrive at the function indicated in the first image.

Do you use the driver with the SPI or MDIO?

**Reply by vChavezB** [reply]
I use the SPI Mode but not with this driver but a custom one I made for Zephyr for the ADIN1110. 

If you are in idle I pressume (perhaps Im wrong) that zephyr already went through main and there is no more work to do. I.e. the CPU is idle. which would mean that the driver is not being initialized at all...

As always do you have an example I can use to compile your project and reproduce your issue ?  (Not only the proj.conf and DTS excerpt you posted here)

**Reply by lmikl** [reply]
I'm also using a custom driver for the ADIN2111. I did but it is not complete. And it works well. I have no problem with it, only for TLS :). that's why i want to use the driver in zephyr. To try to find the origin of the problem.

For example,
you can modify the dchpv4_client example a bit.
I send you a zip with the modified example project.
[dhcpv4_client.zip](https://github.com/zephyrproject-rtos/zephyr/files/12249181/dhcpv4_client.zip)

**Reply by vChavezB** [reply]
ok I will check it out. For which board should I compile it?  From the files I pressume its `nucleo_f207zg`?

i.e. `west build --board=nucloe_f207zg `

**Reply by lmikl** [reply]
it does on any board.
I have several boards. a gigadevice, nucleos. it's the same result on all of them.

i use this command : west build -p always -b nucleo_f207zg but i think it's same

**Reply by vChavezB** [reply]
Ok I just realized the overlay is not in the boards directory. If I recall correctly the overalys should be in boards

i.e.

Your app folder
- proj.conf
- boards (directory
    - your_overlay.overlay

**Reply by vChavezB** [reply]
Please check https://docs.zephyrproject.org/latest/build/dts/howtos.html#set-devicetree-overlays

**Reply by vChavezB** [reply]
Do you get this in your log when Cmake starts?

`Found devicetree overlay: /YOUR_PROJECT_DIR/nucleo_f207zg.overlay`

**Reply by lmikl** [reply]
yes

-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/eclipse/Projects/dhcpv4_client
-- CMake version: 3.27.1
-- Found Python3: /home/eclipse/zephyrproject/.venv/bin/python3 (found suitable exact version "3.8.10") found components: Interpreter 
-- Cache files will be written to: /home/eclipse/.cache/zephyr
-- Zephyr version: 3.4.0 (/home/eclipse/zephyrproject/zephyr)
-- Found west (found suitable version "0.14.0", minimum required is "0.14.0")
-- Board: nucleo_f207zg
-- Found host-tools: zephyr 0.15.1 (/home/eclipse/zephyr-sdk-0.15.1)
-- Found toolchain: zephyr 0.15.1 (/home/eclipse/zephyr-sdk-0.15.1)
-- Found Dtc: /home/eclipse/zephyr-sdk-0.15.1/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /home/eclipse/zephyrproject/zephyr/boards/arm/nucleo_f207zg/nucleo_f207zg.dts
-- Found devicetree overlay: /home/eclipse/Projects/dhcpv4_client/nucleo_f207zg.overlay
-- Generated zephyr.dts: /home/eclipse/Projects/MCA4/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /home/eclipse/Projects/MCA4/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: /home/eclipse/Projects/MCA4/build/zephyr/dts.cmake

**Reply by vChavezB** [reply]
I checked the generated dts and conf header files in the build folder and seem correct. Not sure why the driver does not initialize. I would activate zephyr log in your proj.conf and enable debug messages for the ethernet driver to enable the log module of the adin driver https://github.com/zephyrproject-rtos/zephyr/blob/d7504ab4744c5552c8f6adf04ee9d81d00c6a628/drivers/ethernet/eth_adin2111.c#L8

**Reply by lmikl** [reply]
I have already activated the logs but nothing appears on the console

**Reply by vChavezB** [reply]
Not even the Zephyr Boot string?

**Reply by lmikl** [reply]
No. only uart:~$

**Reply by vChavezB** [reply]
I dont have a board at the moment to test this, but I pressume the driver is somehow not starting as expected. The only thing I can recommend is verifying what drivers are loaded i.e. set a breakpoint where the kernel starts and try to figure out what is happenning

https://github.com/zephyrproject-rtos/zephyr/blob/0aaedf72058b8d492eb491c703de37256ff15247/kernel/init.c#L525C20-L525C28

**Comment by lmikl** [comment]
I have already activated the logs but nothing appears on the console

**Comment by GeorgeCGV** [comment]
@lmikl hello, you might have to modify the priorities and enable extra sub-systems required by it:

```
# Net
CONFIG_NETWORKING=y
CONFIG_NET_L2_ETHERNET=y
## ADIN
CONFIG_ETH_ADIN2111=y
CONFIG_MDIO=y
CONFIG_MDIO_ADIN2111=y
CONFIG_PHY_GENERIC_MII=n
CONFIG_PHY_ADIN2111=y

CONFIG_ETH_ADIN2111_INIT_PRIORITY=71
CONFIG_MDIO_INIT_PRIORITY=72
CONFIG_PHY_INIT_PRIORITY=73
# default
CONFIG_ETH_INIT_PRIORITY=80
```

Priorities play the major role here. Same goes for SPI init priority, should be higher (_numerically lower_) than ADIN.

**Reply by lmikl** [reply]
thanks, it works now.

But i have another problem.
I try to use the simple DHCPv4_client example with the adin.
But i never get IP address.
I have spy with wireshark and no frame is exchanged between the adin2111 and the dhcp server.

my test setup : 
i have a custom board with an GD32F427 and an adin2111.
I have another board with an adin1100. It's the evaluation board of analog device(EVAL_ADIN1100EBZ). It's only to translate SPE to ethernet.
And a PC with DHCP server.

I have a custom driver (but it's really simple) for the ADIN2111 and the DHCPv4_client example works fine. So it's not a setup problem.

here my prj.conf
```
# Net
CONFIG_NET_L2_ETHERNET=y
## ADIN
CONFIG_ETH_ADIN2111=y
CONFIG_MDIO=y
CONFIG_MDIO_ADIN2111=y
CONFIG_PHY_GENERIC_MII=n
CONFIG_PHY_ADIN2111=y
CONFIG_ETH_ADIN2111_SPI_CFG0=n

CONFIG_ETH_ADIN2111_INIT_PRIORITY=71
CONFIG_MDIO_INIT_PRIORITY=72
CONFIG_PHY_INIT_PRIORITY=73
# default
CONFIG_ETH_INIT_PRIORITY=80
CONFIG_NET_IF_MAX_IPV4_COUNT=2

```

and the device tree : 

```
&spi1 {
	pinctrl-0 = <&spi1_default>;
	pinctrl-names = "default";
	interrupts = < 0x24 0x0 >;
	cs-gpios = <&gpiob 12 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
	 status = "okay";
	my_adin: adin2111@0{
		compatible = "adi,adin2111";
		reg = <0x0>;
		spi-max-frequency = <10000000>;
		int-gpios = <&gpioe 14 GPIO_ACTIVE_LOW>;
		reset-gpios = <&gpioe 15 GPIO_ACTIVE_LOW>;
		port1 {
			local-mac-address = [ 00 E0 22 FE DA C8 ];
		};
		port2 {
			local-mac-address = [ 00 E0 22 FE DA C9 ];
		};
		mdio: mdio {
			compatible = "adi,adin2111-mdio";
			#address-cells = <1>;
			#size-cells = <0>;
			phy@1 {
			  reg = <0x1>;
			  compatible = "adi,adin2111-phy";
			  status = "okay";
			};
			phy@2 {
			  reg = <0x2>;
			  compatible = "adi,adin2111-phy";
			  status = "okay";
			};
		};
	};
};
```
If you can help me.
thanks

**Reply by GeorgeCGV** [reply]
I didn't use the DHCP sample. Looks like it uses the default iface, which is the 1st port.
Maybe it is better to validate that you can reach your PC from ADIN2111. For example, using ping with enabled ARP.
If that works fine, then the DHCP sample shall be investigated.

**Reply by lmikl** [reply]
i have another error when the ip address is set : 
```
[00:00:00.118,000] <inf> phy_adin2111: PHY 2 ID 283BCA1
[00:00:00.120,000] <inf> phy_adin2111: PHY 2 2.4V mode supported
[00:00:00.122,000] <inf> phy_adin2111: PHY 1 ID 283BCA1
[00:00:00.124,000] <inf> phy_adin2111: PHY 1 2.4V mode supported
*** Booting Zephyr OS build  ***
[00:00:00.125,000] <inf> net_config: Initializing network
[00:00:00.126,000] <inf> net_config: Waiting interface 1 (0x20000de0) to be up...
[00:00:30.127,000] <inf> net_config: IPv4 address: 192.168.0.10
[00:00:30.127,000] <inf> net_config: Running dhcpv4 client...
[00:00:30.127,000] <err> net_config: Timeout while waiting network interface
[00:00:30.127,000] <err> net_config: Network initialization failed (-115)
[00:00:30.127,000] <inf> net_dhcpv4_client_sample: Run dhcpv4 client
[00:00:30.127,000] <inf> net_dhcpv4_client_sample: adin found
```
I set the ip address with the command
```
CONFIG_NET_CONFIG_SETTINGS=y
CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.168.0.10"
```

ping no works on each port
I may have initialized some things incorrectly.

**Reply by lmikl** [reply]
@GeorgeCGV , I solved the small problem of the timeout but I confirm, the ping doesn't work.

```
[00:00:00.118,000] <inf> phy_adin2111: PHY 2 ID 283BCA1
[00:00:00.121,000] <inf> phy_adin2111: PHY 2 2.4V mode supported
[00:00:00.123,000] <inf> phy_adin2111: PHY 1 ID 283BCA1
[00:00:00.126,000] <inf> phy_adin2111: PHY 1 2.4V mode supported
*** Booting Zephyr OS build  ***
[00:00:00.128,000] <inf> net_config: Initializing network
[00:00:00.128,000] <inf> net_config: Waiting interface 1 (0x20000de0) to be up...
[00:00:00.354,000] <inf> net_config: Interface 1 (0x20000de0) coming up
[00:00:00.354,000] <inf> net_config: IPv4 address: 192.168.0.10
[00:00:00.354,000] <inf> net_config: Running dhcpv4 client...
[00:00:00.354,000] <inf> net_dhcpv4_client_sample: Run dhcpv4 client
[00:00:00.354,000] <inf> net_dhcpv4_client_sample: adin found
uart:~$ net ping 192.168.0.1
PING 192.168.0.1
Ping timeout
uart:~$ 

```

**Reply by GeorgeCGV** [reply]
did you update to get #61596?

does https://github.com/zephyrproject-rtos/zephyr/blob/b28b6a06870f0fc43bb22d7ce2e601ddd69e2b98/drivers/ethernet/eth_adin2111.c#L400 goes on without errors?

what about reception (_ping the device from host_)?
Do you see https://github.com/zephyrproject-rtos/zephyr/blob/b28b6a06870f0fc43bb22d7ce2e601ddd69e2b98/drivers/ethernet/eth_adin2111.c#L152 call (_when host transmits_)?

**Reply by lmikl** [reply]
ok, new update.
the ping works.
But there is a malformed packet for the DHCP request

![MicrosoftTeams-image](https://github.com/zephyrproject-rtos/zephyr/assets/127130921/7c5344a1-3c42-43d6-ad1b-e2e0e897c901)

the frame in text
```

0000   ff ff ff ff ff ff 00 e0 22 fe da c8 08 00 45 00
0010   01 15 00 00 00 00 ff 11 ba d8 00 00 00 00 ff ff
0020   ff ff 00 44 00 43 01 01 05 77 01 01 06 00 98 28
0030   e7 f7 00 00 80 00 00 00 00 00 00 00 00 00 00 00
0040   00 00 00 00 00 00 00 e0 22 fe da c8 00 00 00 00
0050   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0060   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0070   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0080   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0090   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00a0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00b0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00c0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00d0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00e0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00f0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0100   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0110   00 00 00 00 00 00 63 82 53 63 35 01 01 37 03
```

**Reply by lmikl** [reply]
ping doesn't work form the host.
there is a checksum error.

![MicrosoftTeams-image](https://github.com/zephyrproject-rtos/zephyr/assets/127130921/56bbae66-7ce7-48e8-94f8-197cc75790f3)

**Reply by lmikl** [reply]
after a quick debug, I think 4 bytes are missing from each frame send by the driver

**Reply by lmikl** [reply]
i think, i found the error. When the size is transmit by the function eth_adin2111_reg_write the wrong value is transmit. padded_size is transmit instead of burst_size.

but i need to add also 1 to have the end of frame but i don't know why.

ret = eth_adin2111_reg_write(adin, ADIN2111_TX_FSIZE, burst_size+1);

**Reply by GeorgeCGV** [reply]
`ADIN2111_TX_FSIZE` is the frame size, not the spi transfer size. Therefore, burst size is not written there.

**Reply by lmikl** [reply]
yes you are right.

But it missing 4 bytes on each frame send by the adin.
